File: backend/app/schemas/location.py
from datetime import datetime

from pydantic import BaseModel, Field, model_validator

from app.models.location import LocationType


class LocationBase(BaseModel):
    location_name: str = Field(min_length=1, max_length=255)
    location_type: LocationType
    latitude: float | None = Field(default=None, ge=-90, le=90)
    longitude: float | None = Field(default=None, ge=-180, le=180)
    parent_location_id: str | None = None

    @model_validator(mode="after")
    def coordinates_must_be_pair(self) -> "LocationBase":
        if (self.latitude is None) ^ (self.longitude is None):
            raise ValueError("latitude and longitude must be provided together")
        return self


class LocationCreate(LocationBase):
    pass


class LocationUpdate(BaseModel):
    location_name: str | None = Field(default=None, min_length=1, max_length=255)
    location_type: LocationType | None = None
    latitude: float | None = Field(default=None, ge=-90, le=90)
    longitude: float | None = Field(default=None, ge=-180, le=180)
    parent_location_id: str | None = None


class LocationResponse(LocationBase):
    id: str
    created_at: datetime

    model_config = {"from_attributes": True}


class LocationListResponse(BaseModel):
    items: list[LocationResponse]
    total: int
    page: int
    page_size: int

File: frontend/lib/src/features/locations/domain/location.dart
class NetworkLocation {
  const NetworkLocation({
    required this.id,
    required this.locationName,
    required this.locationType,
    required this.createdAt,
    this.latitude,
    this.longitude,
    this.parentLocationId,
  });

  final String id;
  final String locationName;
  final String locationType;
  final double? latitude;
  final double? longitude;
  final String? parentLocationId;
  final DateTime createdAt;

  factory NetworkLocation.fromJson(Map<String, dynamic> json) {
    return NetworkLocation(
      id: json['id'] as String,
      locationName: json['location_name'] as String,
      locationType: json['location_type'] as String,
      latitude: (json['latitude'] as num?)?.toDouble(),
      longitude: (json['longitude'] as num?)?.toDouble(),
      parentLocationId: json['parent_location_id'] as String?,
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }
}

class LocationListResult {
  const LocationListResult({
    required this.items,
    required this.total,
    required this.page,
    required this.pageSize,
  });

  final List<NetworkLocation> items;
  final int total;
  final int page;
  final int pageSize;

  factory LocationListResult.fromJson(Map<String, dynamic> json) {
    return LocationListResult(
      items: (json['items'] as List<dynamic>)
          .map((item) => NetworkLocation.fromJson(item as Map<String, dynamic>))
          .toList(),
      total: json['total'] as int,
      page: json['page'] as int,
      pageSize: json['page_size'] as int,
    );
  }
}
